contracts/exchange-forwarder/contracts/src/MixinExchangeWrapper.sol

```diff
@@ -172,91 +210,43 @@ contract MixinExchangeWrapper is
                 remainingTakerAssetFillAmount,
                 signatures[i]
             );
-
-            // Update amounts filled and fees paid by maker and taker
-            _addFillResults(totalFillResults, singleFillResults);
-
-            // Stop execution if the entire amount of makerAsset has been bought
-            makerAssetFilledAmount = totalFillResults.makerAssetFilledAmount;
-            if (makerAssetFilledAmount >= makerAssetFillAmount) {
-                break;
+            // Percentage fee
+            if (orders[i].makerAssetData.equals(orders[i].takerFeeAssetData)) {
+                // Subtract fee from makerAssetFilledAmount for the net amount acquired.
+                makerAssetAcquiredAmount = _safeAdd(
+                    makerAssetAcquiredAmount,
+                    _safeSub(singleFillResults.makerAssetFilledAmount, singleFillResults.takerFeePaid)
+                );
+
+                wethSpentAmount = _safeAdd(
+                    wethSpentAmount,
+                    singleFillResults.takerAssetFilledAmount
+                );
+            // WETH fee
+            } else {
+                makerAssetAcquiredAmount = _safeAdd(
+                    makerAssetAcquiredAmount,
+                    singleFillResults.makerAssetFilledAmount
+                );
+
+                // WETH is also spent on the taker fee, so we add it here.
+                wethSpentAmount = _safeAdd(
+                    wethSpentAmount,
+                    _safeAdd(singleFillResults.takerAssetFilledAmount, singleFillResults.takerFeePaid)
+                );
```

Comment:
Nit - this for loop and the one above are getting quite long with lots of nesting. I think it could improve readability to break these up.